lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,600 @@
+	    NULL, &attr->nsid, 4);
+	SA_ADD_BULK_ATTR(sa_attrs, (*cnt), sa_tbl[UZFS_ATIME],
+	    NULL, &attr->atime, 16);
+	SA_ADD_BULK_ATTR(sa_attrs, (*cnt), sa_tbl[UZFS_MTIME],
+	    NULL, &attr->mtime, 16);
+	SA_ADD_BULK_ATTR(sa_attrs, (*cnt), sa_tbl[UZFS_CTIME],
+	    NULL, &attr->ctime, 16);
+	SA_ADD_BULK_ATTR(sa_attrs, (*cnt), sa_tbl[UZFS_BTIME],
+	    NULL, &attr->btime, 16);
+}
+
+int
+libuzfs_object_attr_init(libuzfs_dataset_handle_t *dhp,
+    sa_handle_t *sa_hdl, dmu_tx_t *tx)
+{
+	sa_bulk_attr_t sa_attrs[UZFS_END];
+	int cnt = 0;
+	uzfs_attr_t attr;
+	memset(&attr, 0, sizeof (attr));
+	libuzfs_add_bulk_attr(dhp, sa_attrs, &cnt, &attr);
+	return (sa_replace_all_by_template(sa_hdl, sa_attrs, cnt, tx));
+}
+
+int
+libuzfs_get_xattr_zap_obj(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    uint64_t *xattr_zap_obj)
+{
+	objset_t *os = dhp->os;
+	sa_handle_t *sa_hdl = NULL;
+	int err = sa_handle_get(os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	sa_attr_type_t *sa_tbl = dhp->uzfs_attr_table;
+	err = sa_lookup(sa_hdl, sa_tbl[UZFS_ZXATTR], xattr_zap_obj,
+	    sizeof (*xattr_zap_obj));
+	sa_handle_destroy(sa_hdl);
+	return (err);
+}
+
+int
+libuzfs_inode_getattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    uzfs_attr_t *attr)
+{
+	sa_handle_t *sa_hdl;
+	int err = sa_handle_get(dhp->os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	sa_bulk_attr_t sa_attrs[UZFS_END];
+	int cnt = 0;
+	libuzfs_add_bulk_attr(dhp, sa_attrs, &cnt, attr);
+	err = sa_bulk_lookup(sa_hdl, sa_attrs, cnt);
+
+	sa_handle_destroy(sa_hdl);
+	return (err);
+}
+
+int
+libuzfs_inode_setattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
+    const uzfs_attr_t *attr, uint64_t *txg)
+{
+	sa_handle_t *sa_hdl;
+	objset_t *os = dhp->os;
+	int err = sa_handle_get(os, ino, NULL, SA_HDL_PRIVATE, &sa_hdl);
+	if (err != 0) {
+		return (err);
+	}
+
+	dmu_tx_t *tx = dmu_tx_create(os);
+	dmu_tx_hold_sa(tx, sa_hdl, B_FALSE);
+	if ((err = dmu_tx_assign(tx, TXG_WAIT)) != 0) {
+		dmu_tx_abort(tx);
+		return (err);
+	}
+
+	sa_bulk_attr_t sa_attrs[UZFS_END];
+	int cnt = 0;
+	libuzfs_add_bulk_attr(dhp, sa_attrs, &cnt, (uzfs_attr_t *)attr);
+	VERIFY0(sa_bulk_update(sa_hdl, sa_attrs, cnt, tx));
+
+	sa_handle_destroy(sa_hdl);
+
+	if (txg != NULL) {
+		*txg = tx->tx_txg;
+	}
+
+	dmu_tx_commit(tx);
+
+	return (0);
+}
+
+static int
+libuzfs_get_nvlist_from_handle(const sa_attr_type_t *sa_tbl,
+    nvlist_t **nvl, sa_handle_t *sa_hdl)
+{
+	int xattr_sa_size;
+	int err = sa_size(sa_hdl, sa_tbl[UZFS_XATTR], &xattr_sa_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	char *xattr_sa_data = vmem_alloc(xattr_sa_size, KM_SLEEP);
+	err = sa_lookup(sa_hdl, sa_tbl[UZFS_XATTR],
+	    xattr_sa_data, xattr_sa_size);
+	if (err == 0) {
+		err = nvlist_unpack(xattr_sa_data,
+		    xattr_sa_size, nvl, KM_SLEEP);
+	}
+	vmem_free(xattr_sa_data, xattr_sa_size);
+
+	return (err);
+}
+
+static int
+libuzfs_kvattr_set_nvlist(nvlist_t *nvl, const char *name,
```

Comment:
This function has some tricky side effect, that is, if "set nvlist" fails with EFBIG, it will delete the nv if name exists in sa. 

So "set nvlist" is not accurate, because by convention, if "set" fails, it should not impact on the nvlist. 

With that said,  it looks "update nvlist" is a better name for this function. And also, we should add a comment to document this side effect.

Reply:
ok